File: domain/src/main/java/com/ricamgar/notify/domain/reminder/usecase/MarkReminderAsDone.java
package com.ricamgar.notify.domain.reminder.usecase;

import com.ricamgar.notify.domain.base.usecase.UseCase;
import com.ricamgar.notify.domain.reminder.model.Reminder;
import com.ricamgar.notify.domain.reminder.repository.RemindersRepository;

import javax.inject.Inject;
import javax.inject.Named;

import rx.Observable;
import rx.Scheduler;

public class MarkReminderAsDone extends UseCase {

    private final RemindersRepository remindersRepository;
    private Reminder reminder;

    @Inject
    protected MarkReminderAsDone(RemindersRepository remindersRepository,
                                 @Named(value = "ioThread") Scheduler executionThread,
                                 @Named(value = "mainThread") Scheduler postExecutionThread) {
        super(executionThread, postExecutionThread);
        this.remindersRepository = remindersRepository;
    }

    public void setReminder(Reminder reminder) {
        this.reminder = reminder;
    }

    @Override
    protected Observable buildUseCaseObservable() {
        if (reminder == null) {
            return Observable.error(new IllegalArgumentException());
        }

        Reminder doneReminder = new Reminder(reminder.id, reminder.description, reminder.sticky, true);
        return remindersRepository.addOrUpdate(doneReminder);
    }
}

File: domain/src/main/java/com/ricamgar/notify/domain/reminder/usecase/MarkReminderAsDone.kt
package com.ricamgar.notify.domain.reminder.usecase

import com.ricamgar.notify.domain.reminder.model.Reminder
import com.ricamgar.notify.domain.reminder.repository.RemindersRepository
import io.reactivex.Single
import javax.inject.Inject

class MarkReminderAsDone
@Inject constructor(
  private val remindersRepository: RemindersRepository
) {

  fun execute(reminder: Reminder): Single<Reminder> {
    val doneReminder = Reminder(reminder.id, reminder.description, true, reminder.groupId)
    return remindersRepository.addOrUpdate(doneReminder)
  }
}
